File: backend/models/match.go
package models

import (
	"fmt"
	"math"
	"time"
)

// MatchStatus represents the current status of a match
type MatchStatus string

const (
	MatchStatusScheduled MatchStatus = "scheduled"
	MatchStatusFinished  MatchStatus = "finished"
	MatchStatusStarted   MatchStatus = "in-progress"
)

// Match represents a football match with teams, scores and odds
type Match interface {
	Id() string
	GetSeasonCode() string
	GetCompetitionCode() string
	GetDate() time.Time
	GetHomeTeam() string
	GetAwayTeam() string
	GetHomeGoals() int
	GetAwayGoals() int
	GetHomeTeamOdds() float64
	GetAwayTeamOdds() float64
	GetDrawOdds() float64
	SetHomeTeamOdds(odds float64)
	SetAwayTeamOdds(odds float64)
	SetDrawOdds(odds float64)
	AbsoluteGoalDifference() int
	IsDraw() bool
	TotalGoals() int
	AbsoluteDifferenceOddsBetweenHomeAndAway() float64
	GetWinner() string
	// Start marks the match as started
	Start()
	IsFinished() bool
	Finish(homeGoals, awayGoals int)
	IsInProgress() bool
	GetStatus() MatchStatus
}

// SeasonMatch represents a football match within a championship, like Ligue 1, Serie A, etc.
type SeasonMatch struct {
	HomeTeam        string      `json:"homeTeam"`
	AwayTeam        string      `json:"awayTeam"`
	HomeGoals       int         `json:"homeGoals"`
	AwayGoals       int         `json:"awayGoals"`
	HomeTeamOdds    float64     `json:"homeTeamOdds"`
	AwayTeamOdds    float64     `json:"awayTeamOdds"`
	DrawOdds        float64     `json:"drawOdds"`
	Status          MatchStatus `json:"status"`
	SeasonCode      string      `json:"seasonCode"`
	CompetitionCode string      `json:"competitionCode"`
	Date            time.Time   `json:"date"`
	Matchday        int         `json:"matchday"`
}

// NewMatch creates a new Match instance
func NewSeasonMatch(homeTeam, awayTeam string, seasonCode, competitionCode string, date time.Time, matchday int) *SeasonMatch {
	return &SeasonMatch{
		HomeTeam:        homeTeam,
		AwayTeam:        awayTeam,
		Status:          MatchStatusScheduled,
		SeasonCode:      seasonCode,
		CompetitionCode: competitionCode,
		Date:            date,
		Matchday:        matchday,
	}
}

func NewSeasonMatchWithKnownOdds(homeTeam, awayTeam string, seasonCode, competitionCode string, date time.Time, matchday int, homeTeamOdds, awayTeamOdds, drawOdds float64) *SeasonMatch {
	return &SeasonMatch{
		HomeTeam:        homeTeam,
		AwayTeam:        awayTeam,
		HomeTeamOdds:    homeTeamOdds,
		AwayTeamOdds:    awayTeamOdds,
		DrawOdds:        drawOdds,
		Status:          MatchStatusScheduled,
		SeasonCode:      seasonCode,
		CompetitionCode: competitionCode,
		Date:            date,
		Matchday:        matchday,
	}
}

func NewFinishedSeasonMatch(homeTeam, awayTeam string, homeGoals, awayGoals int, seasonCode, competitionCode string, date time.Time, matchday int, homeTeamOdds, awayTeamOdds, drawOdds float64) *SeasonMatch {
	return &SeasonMatch{
		HomeTeam:        homeTeam,
		AwayTeam:        awayTeam,
		HomeTeamOdds:    homeTeamOdds,
		AwayTeamOdds:    awayTeamOdds,
		DrawOdds:        drawOdds,
		HomeGoals:       homeGoals,
		AwayGoals:       awayGoals,
		Status:          MatchStatusFinished,
		SeasonCode:      seasonCode,
		CompetitionCode: competitionCode,
		Date:            date,
		Matchday:        matchday,
	}
}

// GetWinner returns the winning team or "Draw" if the match is tied
func (m *SeasonMatch) GetWinner() string {
	if m.HomeGoals > m.AwayGoals {
		return m.HomeTeam
	}
	if m.AwayGoals > m.HomeGoals {
		return m.AwayTeam
	}
	return "Draw"
}

func (m *SeasonMatch) AbsoluteGoalDifference() int {
	return int(math.Abs(float64(m.HomeGoals - m.AwayGoals)))
}

func (m *SeasonMatch) IsDraw() bool {
	return m.HomeGoals == m.AwayGoals
}

func (m *SeasonMatch) TotalGoals() int {
	return m.HomeGoals + m.AwayGoals
}

func (m *SeasonMatch) AbsoluteDifferenceOddsBetweenHomeAndAway() float64 {
	return math.Abs(m.HomeTeamOdds - m.AwayTeamOdds)
}

func (m *SeasonMatch) IsFinished() bool {
	return m.Status == MatchStatusFinished
}

func (m *SeasonMatch) IsInProgress() bool {
	return m.Status == MatchStatusStarted
}

// Finish marks the match as finished and sets the final score
func (m *SeasonMatch) Finish(homeGoals, awayGoals int) {
	m.HomeGoals = homeGoals
	m.AwayGoals = awayGoals
	m.Status = MatchStatusFinished
}

func (m *SeasonMatch) Id() string {
	return fmt.Sprintf("%s-%s-%s-%s-%d", m.CompetitionCode, m.SeasonCode, m.HomeTeam, m.AwayTeam, m.Matchday)
}

func (m *SeasonMatch) GetSeasonCode() string {
	return m.SeasonCode
}

func (m *SeasonMatch) GetCompetitionCode() string {
	return m.CompetitionCode
}

func (m *SeasonMatch) GetDate() time.Time {
	return m.Date
}

func (m *SeasonMatch) GetHomeTeam() string {
	return m.HomeTeam
}

func (m *SeasonMatch) GetAwayTeam() string {
	return m.AwayTeam
}

func (m *SeasonMatch) GetHomeGoals() int {
	return m.HomeGoals
}

func (m *SeasonMatch) GetAwayGoals() int {
	return m.AwayGoals
}

func (m *SeasonMatch) GetHomeTeamOdds() float64 {
	return m.HomeTeamOdds
}

func (m *SeasonMatch) GetAwayTeamOdds() float64 {
	return m.AwayTeamOdds
}

func (m *SeasonMatch) GetDrawOdds() float64 {
	return m.DrawOdds
}

func (m *SeasonMatch) SetHomeTeamOdds(odds float64) {
	m.HomeTeamOdds = odds
}

func (m *SeasonMatch) SetAwayTeamOdds(odds float64) {
	m.AwayTeamOdds = odds
}

func (m *SeasonMatch) SetDrawOdds(odds float64) {
	m.DrawOdds = odds
}

func (m *SeasonMatch) Start() {
	m.Status = MatchStatusStarted
}

func (m *SeasonMatch) GetStatus() MatchStatus {
	return m.Status
}

File: frontend/ligain/src/types/match.ts
import { Bet } from './bet';

export type MatchStatus = 'scheduled' | 'in-progress' | 'finished';

// Team name mapping for display purposes
const TEAM_NAME_MAPPING: { [key: string]: string } = {
    'Paris': 'Paris FC',
    'Paris Saint Germain': 'PSG',
    'Olympique Marseille': 'Marseille',
    'Olympique Lyonnais': 'Lyon',
    'LOSC Lille': 'Lille',
    'Angers SCO': 'Angers',
};

function getDisplayTeamName(teamName: string): string {
    return TEAM_NAME_MAPPING[teamName] || teamName;
}

export class SeasonMatch {
    private homeTeam: string;
    private awayTeam: string;
    private homeGoals: number;
    private awayGoals: number;
    private homeTeamOdds: number;
    private awayTeamOdds: number;
    private drawOdds: number;
    private status: MatchStatus;
    private seasonCode: string;
    private competitionCode: string;
    private date: Date;
    private matchday: number;

    constructor(
        homeTeam: string,
        awayTeam: string,
        homeGoals: number,
        awayGoals: number,
        homeTeamOdds: number,
        awayTeamOdds: number,
        drawOdds: number,
        status: MatchStatus,
        seasonCode: string,
        competitionCode: string,
        date: string | Date,
        matchday: number
    ) {
        this.homeTeam = homeTeam;
        this.awayTeam = awayTeam;
        this.homeGoals = homeGoals;
        this.awayGoals = awayGoals;
        this.homeTeamOdds = homeTeamOdds;
        this.awayTeamOdds = awayTeamOdds;
        this.drawOdds = drawOdds;
        this.status = status;
        this.seasonCode = seasonCode;
        this.competitionCode = competitionCode;
        this.date = typeof date === 'string' ? new Date(date) : date;
        this.matchday = matchday;
    }

    id(): string {
        return `${this.competitionCode}-${this.seasonCode}-${this.homeTeam}-${this.awayTeam}-${this.matchday}`;
    }

    getHomeTeam(): string {
        return getDisplayTeamName(this.homeTeam);
    }

    getAwayTeam(): string {
        return getDisplayTeamName(this.awayTeam);
    }

    getHomeTeamForLogo(): string {
        return this.homeTeam;
    }

    getAwayTeamForLogo(): string {
        return this.awayTeam;
    }

    getHomeGoals(): number {
        return this.homeGoals;
    }

    getAwayGoals(): number {
        return this.awayGoals;
    }

    getHomeTeamOdds(): number {
        return this.homeTeamOdds;
    }

    getAwayTeamOdds(): number {
        return this.awayTeamOdds;
    }

    getDrawOdds(): number {
        return this.drawOdds;
    }

    getDate(): Date {
        return this.date;
    }

    isFinished(): boolean {
        return this.status === 'finished';
    }

    isInProgress(): boolean {
        return this.status === 'in-progress';
    }

    getWinner(): string {
        if (this.homeGoals > this.awayGoals) {
            return this.getHomeTeam();
        }
        if (this.awayGoals > this.homeGoals) {
            return this.getAwayTeam();
        }
        return 'Draw';
    }

    absoluteGoalDifference(): number {
        return Math.abs(this.homeGoals - this.awayGoals);
    }

    isDraw(): boolean {
        return this.homeGoals === this.awayGoals;
    }

    totalGoals(): number {
        return this.homeGoals + this.awayGoals;
    }

    absoluteDifferenceOddsBetweenHomeAndAway(): number {
        return Math.abs(this.homeTeamOdds - this.awayTeamOdds);
    }

    hasClearFavorite(): boolean {
        return this.absoluteDifferenceOddsBetweenHomeAndAway() > 1.5;
    }

    getFavoriteTeam(): string {
        if (!this.hasClearFavorite()) {
            return '';
        }
        if (this.homeTeamOdds < this.awayTeamOdds) {
            return this.getHomeTeam();
        }
        return this.getAwayTeam();
    }

    finish(homeGoals: number, awayGoals: number): void {
        this.homeGoals = homeGoals;
        this.awayGoals = awayGoals;
        this.status = 'finished';
    }

    getSeasonCode(): string {
        return this.seasonCode;
    }

    getCompetitionCode(): string {
        return this.competitionCode;
    }

    getMatchday(): number {
        return this.matchday;
    }

    hasStarted(now: Date): boolean {
        return now >= this.date;
    }

    static fromJSON(json: any): SeasonMatch {
        return new SeasonMatch(
            json.homeTeam,
            json.awayTeam,
            json.homeGoals,
            json.awayGoals,
            json.homeTeamOdds,
            json.awayTeamOdds,
            json.drawOdds,
            json.status,
            json.seasonCode,
            json.competitionCode,
            json.date,
            json.matchday
        );
    }
}

export interface SimplifiedBet {
    playerId: string;
    playerName: string;
    predictedHomeGoals: number;
    predictedAwayGoals: number;
    isModifiable(now: Date): boolean;
}

export interface SimplifiedScore {
    playerId: string;
    playerName: string;
    points: number;
}

export interface MatchResult {
    match: SeasonMatch;
    bets: { [key: string]: SimplifiedBet } | null;
    scores: { [key: string]: SimplifiedScore } | null;
}

export interface MatchesResponse {
    incomingMatches: { [key: string]: MatchResult };
    pastMatches: { [key: string]: MatchResult };
} 